pkg/scheduler/queue/tree/tenant_querier_queuing_algorithm.go

```diff
@@ -222,27 +224,43 @@ func (qa *TenantQuerierQueuingAlgorithm) dequeueUpdateState(node *Node, dequeued
 	// delete from shared tenantNodes
 	for i, tenantNode := range qa.tenantNodes[childName] {
 		if tenantNode == dequeuedFrom {
-			qa.tenantNodes[childName] = append(qa.tenantNodes[childName][:i], qa.tenantNodes[childName][i+1:]...)
+			qa.tenantNodes[childName] = slices.Delete(qa.tenantNodes[childName], i, i+1)
+
 		}
-	}
 
-	// check tenantNodes; we only remove from tenantIDOrder if all nodes with this name are gone.
-	// If the removed child is at the _end_ of tenantIDOrder, we remove it outright; otherwise,
-	// we replace it with an empty string. Removal of elements from anywhere other than the end of the slice
-	// would re-index all tenant IDs, resulting in skipped tenants when starting iteration from the
-	// querier-provided LastTenantIndex.
-	removeFromSharedQueueOrder := len(qa.tenantNodes[childName]) == 0
-
-	if removeFromSharedQueueOrder {
-		for idx, name := range qa.tenantIDOrder {
-			if name == childName {
-				qa.tenantIDOrder[idx] = string(emptyTenantID)
+		// check tenantNodes; we only remove from tenantIDOrder if all nodes with this name are gone.
+		removeFromSharedQueueOrder := len(qa.tenantNodes[childName]) == 0
+
+		// When tenantIDOrder is very long, removing tenants from the order is performance-sensitive
+		// when clearing & deleting many tenant queues at the same time due to rollout operations.
+		if removeFromSharedQueueOrder {
+
+			// First, just replace the tenantID with the emptyTenantID sentinel value.
+			for idx, name := range qa.tenantIDOrder {
+				if name == childName {
+					qa.tenantIDOrder[idx] = string(emptyTenantID)
+					// tenant ID should only appear once in the list
+					break
+				}
 			}
-		}
-		// clear all sequential empty elements from tenantIDOrder
-		lastElementIndex := len(qa.tenantIDOrder) - 1
-		for i := lastElementIndex; i >= 0 && qa.tenantIDOrder[i] == ""; i-- {
-			qa.tenantIDOrder = qa.tenantIDOrder[:i]
+
+			// Then, only shrink tenantIDOrder when we can remove emptyTenantID sentinel values from the end.
+			// Shrinking from the middle would re-index the tenant order and could cause us to skip tenants
+			// when starting iteration from the LastTenantIndex provided by a querier-worker.
+			//
+			// Because we wait to shrink tenantIDOrder until we can remove the last element from the end,
+			// there may be many consecutive emptyTenantIDs up to (last index - 1) waiting to be removed;
+			// count up the emptyTenantIDs at the end and remove all at once to avoid multiple slice resizes.
+			emptyTenantIDsAtEnd := 0
+			for i := len(qa.tenantIDOrder) - 1; i >= 0 && qa.tenantIDOrder[i] == ""; i-- {
+				emptyTenantIDsAtEnd++
+			}
+
+			qa.tenantIDOrder = slices.Delete(
+				qa.tenantIDOrder,
+				len(qa.tenantIDOrder)-emptyTenantIDsAtEnd,
+				len(qa.tenantIDOrder),
```

Comment:
Is there a reason why you're using `slices.Delete` here rather than reslicing with something like `qa.tenantIDOrder = qa.tenantIDOrder[:len(qa.tenantIDOrder)-emptyTenantIDsAtEnd]`? I'd expect that to be a little faster given we don't need some of the things that `slices.Delete` does like clearing out the deleted elements.

Reply:
`slices.Delete` is just so much easier to understand when reading.
Re-benchmarking those two options shows:
* not using slices.Delete gets 61.53% speedup
* using it gets 61.06% speedup

I am not quite convinced that is a point of optimization we need here.

Reply:
If the concern is readability, what if we introduce a `deleteFromEndOfSlice` method (with a better name than that 🙂)?

Reply:
@charleskorn I have introduced this PR in order to accomplish this - essentially copying over the slices utils from Go stdlib as they were in 1.21 before the clearing of elements was introduced for the deletion-related utils

needs a rubber-stamp, but I got the go-ahead on the license header setup from Bryan Boreham https://github.com/grafana/mimir/pull/9894

Reply:
if adding that package is going to be a big deal I would prefer to just merge this as it is as the hot path is benchmarked.